File: c++/memkvstore.cc
#include <glog/logging.h>
#include <optional>
#include <string>

#include "memkvstore.h"

namespace kvstore {

using multipaxos::Command;

std::optional<std::string> MemKVStore::Get(std::string const& key) {
  auto it = map_.find(key);
  if (it != map_.end())
    return it->second;
  return std::nullopt;
}

bool MemKVStore::Put(std::string const& key, std::string const& value) {
  map_[key] = value;
  return true;
}

bool MemKVStore::Del(std::string const& key) {
  return map_.erase(key) != 0;
}

KVResult MemKVStore::Execute(Command const& cmd) {
  if (cmd.type() == multipaxos::CommandType::GET) {
    auto r = Get(cmd.key());
    if (!r)
      return KVResult{false, kKeyNotFound};
    return KVResult{true, std::move(*r)};
  }

  if (cmd.type() == multipaxos::CommandType::PUT) {
    Put(cmd.key(), cmd.value());
    return KVResult{true, kEmpty};
  }

  CHECK(cmd.type() == multipaxos::CommandType::DEL);

  if (Del(cmd.key()))
    return KVResult{true, kEmpty};
  return KVResult{false, kKeyNotFound};
}

}  // namespace kvstore

File: java-tcp/replicant/src/main/java/kvstore/MemKVStore.java
package kvstore;

import command.Command;
import command.Command.CommandType;
import command.KVResult;

import java.util.HashMap;
public class MemKVStore implements KVStore {

  public static final String kKeyNotFound = "key not found";
  public static final String kEmpty = "";
  private final HashMap<String, String> store;

  public MemKVStore() {
    this.store = new HashMap<>();
  }

  public String get(String key) {
    return store.get(key);
  }

  public boolean put(String key, String value) {
    store.put(key, value);
    return true;
  }

  public boolean del(String key) {
    String value = store.remove(key);
    return value != null;
  }

  @Override
  public KVResult execute(Command cmd) {
    CommandType cmdType = cmd.getCommandType();
    if (cmdType == CommandType.Get) {
      String r = get(cmd.getKey());
      if (r == null) {
        return new KVResult(false, kKeyNotFound);
      } else {
        return new KVResult(true, r);
      }
    } else if (cmdType == CommandType.Put) {
      put(cmd.getKey(), cmd.getValue());
      return new KVResult(true, kEmpty);
    } else if (cmdType == CommandType.Del && del(cmd.getKey())) {
      return new KVResult(true, kEmpty);
    } else {
      return new KVResult(false, kKeyNotFound);
    }
  }
}

